tests/jax/test_custom_partitioning.py

```diff
@@ -0,0 +1,301 @@
+        devices = np.asarray(jax.devices()[:DEVICE_COUNT]).reshape(*mesh_shape)
+        with global_shard_guard(_get_sharding_resource(mesh_names, sharding_type)[0]):
+            with jax.sharding.Mesh(devices, mesh_names):
+                x_ = random.normal(random.PRNGKey(1124), input_shape)
+                mask_ = make_attention_mask(tokens)
+                graded_f = jax.value_and_grad(func, argnums=(0))
+                pjitter = pjit(graded_f)
+                out = {}
+                hlo = pjitter.lower(x_, mask_).compile().as_text()
+                dic = count_collective(hlo)
+                assert dic==collective_ref, f"Expected number of collective is: {dic==collective_ref=}, but got {dic=}."
+
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
+    @pytest.mark.parametrize('input_shape', [(32, 128, 16, 64)])
+    @pytest.mark.parametrize('pad_ratio', [0.3])
+    @pytest.mark.parametrize('dropout_rng', [0.1])
+    @pytest.mark.parametrize('attn_bias_type', [AttnBiasType.NO_BIAS, AttnBiasType.POST_SCALE_BIAS])
+    @pytest.mark.parametrize('attn_mask_type', [AttnMaskType.PADDING_MASK, AttnMaskType.CAUSAL_MASK])
+    def test_fused_attn(self, mesh_shape, mesh_names, sharding_type, collective_ref,
+                        input_shape, pad_ratio, dropout_probability, attn_bias_type, attn_mask_type):
+        b, s, h, d = input_shape
+        scaling_factor = 1. / jnp.sqrt(d)
+        key = jax.random.PRNGKey(1124)
+        subkeys = jax.random.split(key, 3)
+        
+        def func(x, bias, mask):
+            x = self_fused_attn(x, bias, mask, subkeys[2], attn_bias_type, attn_mask_type,
+                                scaling_factor, dropout_probability, True,
+                                sharding_type=sharding_type)
+            return jnp.mean(x)
+
+        qkv_shape = (b, s, 3, h, d)
+        bias_shape = (1, h, s, s)
+        pad_len = int(s * pad_ratio)
+        valid_len = s - pad_len
+        tokens = jnp.concatenate((jnp.ones((b, valid_len)), jnp.zeros((b, pad_len))), axis=-1)
+        devices = np.asarray(jax.devices()[:DEVICE_COUNT]).reshape(*mesh_shape)
+        with global_shard_guard(_get_sharding_resource(mesh_names, sharding_type)[0]):
+            with jax.sharding.Mesh(devices, mesh_names):
+                x_ = random.normal(subkeys[0], qkv_shape)
+                bias_ = random.normal(subkeys[1], bias_shape)
+                mask_ = make_decoder_mask(tokens)
+                graded_f = jax.value_and_grad(func, argnums=(0, 1, 2))
+                pjitter = pjit(graded_f)
+                out = {}
+                hlo = pjitter.lower(x_, bias_, mask_).compile().as_text()
+                dic = count_collective(hlo)
+                assert dic==collective_ref, f"Expected number of collective is: {dic==collective_ref=}, but got {dic=}."
+
+class TestCPGenerator:
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
```

Comment:
Regarding softmax, it is expected 0 collective ops for all sharding. Since no weights need to reduce.